sw/device/lib/dif/dif_padctrl.h

```diff
@@ -0,0 +1,196 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_DIF_PADCTRL_H_
+#define OPENTITAN_SW_DEVICE_LIB_DIF_PADCTRL_H_
+
+#include "sw/device/lib/base/mmio.h"
+
+// TODO - Should be autogenerated.
+// - Should padctrl/pinmux share some defines/constants (such as MIO_NUM) or do
+// we just autogen the same number twice?
+#define DIF_PADCTRL_DIO_NUM 32
+#define DIF_PADCTRL_MIO_NUM 32
+#define DIF_PADCTRL_ATTR_WIDTH 8
+#define DIF_PADCTRL_ADDITIONAL_ATTR_WIDTH 2
+
+/**
+ * Pad attributes
+ *
+ * Some attributes have defined meanings, others are implementation defined
+ * (|kDifPadCtrlAttrAdditionalFirst| and onwards). not all pads will support all
+ * attributes.
+ */
+typedef enum dif_padctrl_attr {
+  kDifPadctrlAttrIOInvert = 0,
+  kDifPadctrlAttrOpenDrain,
+  kDifPadctrlAttrPullDown,
+  kDifPadctrlAttrPullUp,
+  kDifPadctrlAttrKeeper,
+  kDifPadctrlAttrWeakDrive
+  kDifPadctrlAttrAdditionalFirst = 0,
+  kDifPadctrlAttrAdditionalLast =
+    DIF_PADCTRL_ADDITIONAL_ATTR_WIDTH == 0 ?
+      kDifPadctrlAdditionalAttrFirst :
+      kDifPadctrlAdditionalAttrFirst + DIF_PADCTRL_ADDITIONAL_ATTR_WIDTH - 1,
+} dif_padctrl_attr_t;
+
+/**
+ * DIO pads
+ */
+typedef enum dif_padctrl_dio {
+  kDifPadctrlDioFirst = 0,
+  kDifPadctrlDioLast = DIF_PADCTRL_DIO_NUM - 1,
+} dif_padctrl_dio_t;
+
+/**
+ * MIO Pads
+ */
+typedef enum dif_padctrl_mio {
+  kDifPadctrlMioFirst = 0,
+  kDifPadctrlMioLast = DIF_PADCTRL_MIO_NUM - 1,
+} dif_padctrl_mio_t;
+
+/**
+ * Padctrl instance state.
+ *
+ * Padctrl persistent data that is required by the Padctrl API.
+ */
+typedef struct dif_padctrl {
+  mmio_region_t base_addr; /**< Padctrl base address. */
+  bool write_enabled;      /**< Configuration registers locked/unlocked. */
+} dif_padctrl_t;
+
+/**
+ * Padctrl init routine error codes.
+ */
+typedef enum dif_padctrl_init_error {
+  kDifPadctrlInitErrorNone = 0         /**< Success. */
+  kDifPadctrlInitErrorInvalidArgument, /**< Input parameter is not valid. */
+  kDifPadctrlInitErrorRegistersLocked, /**< Peripheral is locked and cannot be
+                                          reset. */
+} dif_padctrl_init_error_t;
+
+/**
+ * Initialises an instance of Padctrl
```

Comment:
Could you add full stops at the end of every sentences? I always miss them myself, and then Miguel reminds me :)